pkg/controller/volumeattachment.go

```diff
@@ -0,0 +1,253 @@
+		return reconcile.Result{}, err
+	}
+
+	azVolumeName := strings.ToLower(diskName)
+	nodeName := volumeAttachment.Spec.NodeName
+	return r.AnnotateAzVolumeAttachment(ctx, volumeAttachment, azureutils.GetAzVolumeAttachmentName(azVolumeName, nodeName), volumeAttachment.Name, volumeAttachmentExists)
+}
+
+func (r *ReconcileVolumeAttachment) AnnotateAzVolumeAttachment(ctx context.Context, volumeAttachment storagev1.VolumeAttachment, azVolumeAttachmentName, volumeAttachmentName string, volumeAttachmentExists bool) (reconcile.Result, error) {
+	var azVolumeAttachment v1alpha1.AzVolumeAttachment
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: azVolumeAttachmentName}, &azVolumeAttachment); err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to get AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		r.retryMutex.RLock()
+		numRetry, ok := r.retryMap[azVolumeAttachmentName]
+		r.retryMutex.RUnlock()
+		if !ok {
+			r.retryMutex.Lock()
+			if _, ok = r.retryMap[azVolumeAttachmentName]; !ok {
+				r.retryMap[azVolumeAttachmentName] = 0
+				numRetry = 0
+			}
+			r.retryMutex.Unlock()
+		}
+
+		if numRetry <= maxRetry {
+			r.retryMutex.Lock()
+			r.retryMap[azVolumeAttachmentName]++
+			r.retryMutex.Unlock()
+			klog.V(5).Infof("Waiting for AzVolumeAttachment (%s) to be created...", azVolumeAttachmentName)
+			return reconcile.Result{Requeue: true}, nil
+		}
+
+		klog.V(5).Infof("Max Retry (%d) for Get AzVolumeAttachment (%s) exceeded. The CRI is probably deleted.", maxRetry, azVolumeAttachmentName)
+		r.retryMutex.Lock()
+		delete(r.retryMap, azVolumeAttachmentName)
+		r.retryMutex.Unlock()
+		return reconcile.Result{}, nil
+	}
+
+	patched := azVolumeAttachment.DeepCopy()
+	if patched.Annotations == nil {
+		patched.Annotations = make(map[string]string)
+	}
+
+	_, ok := patched.Annotations[azureutils.VolumeAttachmentExistsAnnotation]
+	if volumeAttachmentExists {
+		if ok {
+			return reconcile.Result{}, nil
+		}
+		patched.Annotations[azureutils.VolumeAttachmentExistsAnnotation] = volumeAttachmentName
+	} else {
+		if !ok {
+			return reconcile.Result{}, nil
+		}
+		delete(patched.Annotations, azureutils.VolumeAttachmentExistsAnnotation)
+	}
+
+	if err := r.client.Patch(ctx, patched, client.MergeFrom(&azVolumeAttachment)); err != nil {
+		klog.Errorf("failed to update AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+	klog.V(2).Infof("successfully updated AzVolumeAttachment (%s) with annotation (%s)", azVolumeAttachmentName, azureutils.VolumeAttachmentExistsAnnotation)
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileVolumeAttachment) Recover(ctx context.Context) error {
+	// Get all volumeAttachments
+	volumeAttachments, err := r.kubeClient.StorageV1().VolumeAttachments().List(ctx, metav1.ListOptions{})
```

Comment:
Can't we get the volumeAttachments where volumeAttachment.Spec.Attacher == azureutils.DriverName in the first place instead of filtering them below?

Reply:
In order to do so, `spec.Attacher` needs to be a supported field selector (https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#supported-fields), which I don't think is.